src/main/java/it/gov/pagopa/payhub/activities/service/IuvService.java

```diff
@@ -0,0 +1,109 @@
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service class exposing methods related to IUV handling.
+ */
+@Lazy
+@Service
+@Slf4j
+public class IuvService {
+
+  private static final String PROGRESSIVO_PADDING_ZEROES_SMALL = "%011d";
+  public static final String AUX_DIGIT = "3";
+
+  private final String informationSystemId;
+
+  private final IuvSequenceNumberDao progressiviVersamentoDao;
+
+  public IuvService(@Value("${iuv.informationSystemId:00}") String informationSystemId,
+                    IuvSequenceNumberDao iuvSequenceNumberDao) {
+    this.informationSystemId = informationSystemId;
+    this.progressiviVersamentoDao = iuvSequenceNumberDao;
+  }
+
+
+  /**
+   * Generate a valid and unique IUV given the organization entity.
+   * @param org the organization for which to generate the IUV
+   * @return the generated IUV
+   */
+  public String generateIuv(OrganizationDTO org){
+    StringBuilder iuvBuilder = new StringBuilder();
+    iuvBuilder.append(org.getApplicationCode());
+    iuvBuilder.append(informationSystemId);
+    long paymentIndex = progressiviVersamentoDao.getNextIuvSequenceNumber(org.getIpaCode());
+    if(paymentIndex<1){
+      log.error("invalid payment index returned for org[{}/{}]: {}", org.getIpaCode(), org.getOrgFiscalCode(), paymentIndex);
+      throw new ValueNotValidException("invalid payment index");
+    }
+    iuvBuilder.append(String.format(PROGRESSIVO_PADDING_ZEROES_SMALL, paymentIndex));
+
+    //check digit
+    String digitString = AUX_DIGIT + iuvBuilder;
+    long digit = Long.parseLong(digitString);
+    long reminder = digit % 93;
+    String checkDigit = StringUtils.leftPad(String.valueOf(reminder), 2, '0');
+    iuvBuilder.append(checkDigit);
+
+    return iuvBuilder.toString();
+  }
+
+  /**
+   * Utility method to generate the NAV (notice number) given the corresponding IUV.
+   * @param iuv the IUV for which to generate the NAV
+   * @return the generated NAV
+   */
+  public String iuv2Nav(String iuv){
+    if(isValidIuv(iuv))
+      return AUX_DIGIT + iuv;
+    else
+      throw new ValueNotValidException("invalid iuv");
+  }
+
+  /**
+   * Utility method to extract the IUV given the corresponding NAV (notice number).
+   * @param nav the NAV for which to extract the IUV
+   * @return the extraxted IUV
+   */
+  public String nav2Iuv(String nav){
+    if(isValidNav(nav)){
+      return nav.substring(AUX_DIGIT.length());
+    } else {
+      throw new ValueNotValidException("invalid nav");
+    }
+  }
+
+  /**
+   * Utility method to formally validate a IUV.
+   * @param iuv the IUV to validate
+   * @return true if valid, otherwise false
+   */
+  public boolean isValidIuv(String iuv){
+    return isValidNav(StringUtils.join(AUX_DIGIT+iuv));
```

Comment:
the join method is not really used, just use the simple concat operation.
Moreover, you are performing this concatenation also at row 67, why don't you externalize it on a private method `innerIuv2Nav` instead of replicating it?

Reply:
There was a typo that I fixed, the correct join is:
`isValidNav(StringUtils.join(AUX_DIGIT,iuv))`
I would prefer keeping the `join` function since it better handle the null (it will prevent having `"a"+null` resulting as an `"anull"` string).
Also while apparently this is the same operation used inside `iuv2Nav` method:
```
    if(isValidIuv(iuv))
      return AUX_DIGIT + iuv;
```
the context is different in a subtle way: here the concat is used on an already validated iuv so there is no risk of having null, therefore there is no need to use the join.